File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/49862-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom SharePoint Service Applications/VB/WCF_CustomServiceApplication/DayNamerService.vb
Imports Microsoft.SharePoint.Administration
Imports WCF_CustomServiceApplication.Client

Namespace Server
    'This class creates a link to create a new Day Namer service
    'application in the Manage Service Applications page ribbon
    'in Central Administration. It also creates the service application
    'and its proxy.

    <System.Runtime.InteropServices.Guid("32C35AAD-7B0C-49D9-8626-03186D18CB8E")> _
    Public Class DayNamerService
        Inherits SPIisWebService
        Implements IServiceAdministration

#Region "Constructors"

        Public Sub New()

        End Sub

        Public Sub New(ByVal farm As SPFarm)
            MyBase.New(farm)
        End Sub

#End Region

        'This method creates the service application
        Public Function CreateApplication(ByVal name As String, _
                                          ByVal serviceApplicationType As Type, _
                                          ByVal provisioningContext As SPServiceProvisioningContext) _
                                      As SPServiceApplication _
                                      Implements IServiceAdministration.CreateApplication

            'Validation checks
            If Not serviceApplicationType Is GetType(DayNamerServiceApplication) Then
                Throw New NotSupportedException()
            End If
            If provisioningContext Is Nothing Then
                Throw New ArgumentNullException("provisioningContext")
            End If

            'If the service doesn't already exist, create it
            Dim serviceApp As DayNamerServiceApplication = TryCast(Me.Farm.GetObject(name, Me.Id, serviceApplicationType), DayNamerServiceApplication)
            If serviceApp Is Nothing Then
                serviceApp = DayNamerServiceApplication.Create(name, Me, provisioningContext.IisWebServiceApplicationPool)
            End If

            Return serviceApp

        End Function

        Public Function CreateProxy(ByVal name As String, _
                                    ByVal serviceApplication As SPServiceApplication, _
                                    ByVal provisioningContext As SPServiceProvisioningContext) _
                                As SPServiceApplicationProxy _
                                Implements IServiceAdministration.CreateProxy

            'Validation checks
            If Not serviceApplication.GetType() Is GetType(DayNamerServiceApplication) Then
                Throw New NotSupportedException()
            End If
            If serviceApplication Is Nothing Then
                Throw New ArgumentNullException("serviceApplication")
            End If

            'Verify the service proxy exists
            Dim serviceProxy As DayNamerServiceProxy = DirectCast(Me.Farm.GetObject(name, Me.Farm.Id, GetType(DayNamerServiceProxy)), DayNamerServiceProxy)
            If serviceProxy Is Nothing Then
                Throw New InvalidOperationException("DayNamerServiceProxy does not exist in the farm.")
            End If

            'If the app proxy doesn't exist, create it
            Dim applicationProxy As DayNamerServiceApplicationProxy = serviceProxy.ApplicationProxies.GetValue(Of DayNamerServiceApplicationProxy)(name)
            If applicationProxy Is Nothing Then
                Dim serviceAppAddress As Uri = DirectCast(serviceApplication, DayNamerServiceApplication).Uri
                applicationProxy = New DayNamerServiceApplicationProxy(name, serviceProxy, serviceAppAddress)
            End If

            Return applicationProxy

        End Function

        'This method returns a description of the service application
        Public Function GetApplicationTypeDescription(ByVal serviceApplicationType As Type) _
            As SPPersistedTypeDescription _
            Implements IServiceAdministration.GetApplicationTypeDescription

            If Not serviceApplicationType Is GetType(DayNamerServiceApplication) Then
                Throw New NotSupportedException()
            End If
            Return New SPPersistedTypeDescription("Day Namer Service", "Custom service application providing simple date lookups.")
        End Function

        Public Function GetApplicationTypes() _
            As Type() _
            Implements IServiceAdministration.GetApplicationTypes

            Dim returnType As Type() = {GetType(DayNamerServiceApplication)}
            Return returnType

        End Function

        Public Overrides Function GetCreateApplicationLink(ByVal serviceApplicationType As Type) _
            As SPAdministrationLink _
            Implements IServiceAdministration.GetCreateApplicationLink

            'Make sure this links to your own Create.aspx page. 
            Return New SPAdministrationLink("/_admin/DayNamerService/Create.aspx")

        End Function

        Public Overrides Function GetCreateApplicationOptions(ByVal serviceApplicationType As System.Type) _
            As SPCreateApplicationOptions _
            Implements IServiceAdministration.GetCreateApplicationOptions

            Return MyBase.GetCreateApplicationOptions(serviceApplicationType)

        End Function

    End Class

End Namespace

File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/49862-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom SharePoint Service Applications/C#/WCF_CustomServiceApplication/DayNamerService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.SharePoint.Administration;
using WCF_CustomServiceApplication.Client;

namespace WCF_CustomServiceApplication.Server
{
    //This class creates a link to create a new Day Namer service
    //application in the Manage Service Applications page ribbon
    //in Central Administration. It also creates the service application
    //and its proxy.

    [System.Runtime.InteropServices.Guid("F4E0F76F-96C6-49D3-9475-E2127FEBCC68")]
    public class DayNamerService : SPIisWebService, IServiceAdministration
    {
        public DayNamerService() { }
        public DayNamerService(SPFarm farm)
            : base(farm) { }

        //This method creates the service application
        public SPServiceApplication CreateApplication(string name, Type serviceApplicationType, SPServiceProvisioningContext provisioningContext)
        {
            #region validation
            if (serviceApplicationType != typeof(DayNamerServiceApplication))
                throw new NotSupportedException();
            if (provisioningContext == null)
                throw new ArgumentNullException("provisioningContext");
            #endregion

            // if the service doesn't already exist, create it
            DayNamerServiceApplication serviceApp = this.Farm.GetObject(name, this.Id, serviceApplicationType) as DayNamerServiceApplication;
            if (serviceApp == null)
                serviceApp = DayNamerServiceApplication.Create(name, this, provisioningContext.IisWebServiceApplicationPool);

            return serviceApp;
        }

        //This method creates the service application proxy
        public SPServiceApplicationProxy CreateProxy(string name, SPServiceApplication serviceApplication, SPServiceProvisioningContext provisioningContext)
        {
            #region validation
            if (serviceApplication.GetType() != typeof(DayNamerServiceApplication))
                throw new NotSupportedException();
            if (serviceApplication == null)
                throw new ArgumentNullException("serviceApplication");
            #endregion

            // verify the service proxy exists
            DayNamerServiceProxy serviceProxy = (DayNamerServiceProxy)this.Farm.GetObject(name, this.Farm.Id, typeof(DayNamerServiceProxy));
            if (serviceProxy == null)
                throw new InvalidOperationException("DayNamerServiceProxy does not exist in the farm.");

            // if the app proxy doesn't exist, create it
            DayNamerServiceApplicationProxy applicationProxy = serviceProxy.ApplicationProxies.GetValue<DayNamerServiceApplicationProxy>(name);
            if (applicationProxy == null)
            {
                Uri serviceAppAddress = ((DayNamerServiceApplication)serviceApplication).Uri;
                applicationProxy = new DayNamerServiceApplicationProxy(name, serviceProxy, serviceAppAddress);
            }

            return applicationProxy;
        }

        //This method returns a description of the service application
        public SPPersistedTypeDescription GetApplicationTypeDescription(Type serviceApplicationType)
        {
            if (serviceApplicationType != typeof(DayNamerServiceApplication))
                throw new NotSupportedException();

            return new SPPersistedTypeDescription("Day Namer Service", "Custom service application providing simple date lookups.");
        }

        public Type[] GetApplicationTypes()
        {
            return new Type[] { typeof(DayNamerServiceApplication) };
        }

        public override SPAdministrationLink GetCreateApplicationLink(Type serviceApplicationType)
        {
            //Make sure this links to your own Create.aspx page. 
            return new SPAdministrationLink("/_admin/DayNamerService/Create.aspx");
        }
    }
}
